### PR TITLE
Renamed 'Generic Column' -> 'Numeric Column'; Fixed a few resource leaks in processing; misc refinements

### PR DESCRIPTION
**Pull request by leventov**
This PR accumulates many refactorings and small improvements that I did while preparing the next change set of https://github.com/druid-io/druid/projects/2. I finally decided to make them a separate PR to minimize the volume of the main PR.

Some of the changes:
 - Renamed confusing "Generic Column" term to "Numeric Column" (what it actually implies) in many class names.
 - Generified `ComplexMetricExtractor`

### CODE REVIEW

**Comment by leventov** [comment]
@nishantmonu51 FYI I've tried to fix #5956 opportunistically in this PR, but a lot of tests started to fail. So I reverted.

File: processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java [hunk 1]
```diff
@@ -353,6 +354,15 @@ public double aggregate(final double current, final BaseObjectColumnValueSelecto
         return Context.toNumber(res);
       }
 
+      private Object[] arrayToObjectArray(Object array)
+      {
+        final Object[] objectArray = new Object[Array.getLength(array)];
+        for (int j = 0; j < Array.getLength(array); j++) {
```
Comment (asdf2014) [inline processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java]: Maybe this line of code can become this way?

```java
int len = Array.getLength(array);
final Object[] objectArray = new Object[len];
for (int j = 0; j < len; j++) {...}
```

Cuz, `for (int j = 0; j < Array.getLength(array); j++) {...}` this code style will call `Array.getLength` method for each `for` loop.

File: processing/src/main/java/io/druid/collections/spatial/ImmutableRTree.java [hunk 2]
```diff
@@ -75,7 +74,7 @@ public static ImmutableRTree newImmutableFromMutable(RTree rTree)
       return empty();
     }
 
-    ByteBuffer buffer = ByteBuffer.wrap(new byte[calcNumBytes(rTree)]);
+    ByteBuffer buffer = ByteBuffer.allocate(calcNumBytes(rTree));
```
Comment (asdf2014) [inline processing/src/main/java/io/druid/collections/spatial/ImmutableRTree.java]: Maybe using `ByteBuffer.allocateDirect` will get better performance.

```java
ByteBuffer buffer = ByteBuffer.allocateDirect(calcNumBytes(rTree));
```

**Comment by asdf2014** [comment]
Hi, @leventov . Already created a small PR [#5980](https://github.com/apache/incubator-druid/pull/5980) to supplement the current PR. If u think it's necessary, then I will create a series of PRs to make it more complete.

**Comment by nishantmonu51** [comment]
@leventov : can we finish up the review and merge https://github.com/apache/incubator-druid/pull/5958 before making any more major refactorings ?

**Comment by leventov** [comment]
@nishantmonu51 we can (however it will take several more weeks), but you could review this PR in parallel?

**Comment by himanshug** [comment]
> Renamed `GenericColumnSerializer` to just `ColumnSerializer` and changed the signature of it's serialize() method. This class was annotated @ExtensionPoint, however it doesn't look so to me.

It is extensible when someone writes a custom aggreagtor and extension wants to take control of full column serialization instead default behavior where extension writer only controls serialization of individual row values by providing a ObjectStrategy.
We do have few aggregators (not in the open source) that take advantage of it.

File: processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java (hunk 1 above)
Comment (leventov) [reply processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java]: Thanks, extracted a variable.

File: processing/src/test/java/io/druid/collections/bitmap/BitmapBenchmark.java [hunk 3]
```diff
@@ -73,7 +73,7 @@ protected static ImmutableConciseSet makeOffheapConcise(ImmutableConciseSet conc
     conciseCount++;
     final ByteBuffer buf = ByteBuffer.allocateDirect(bytes.length).put(bytes);
     buf.rewind();
-    return new ImmutableConciseSet(buf);
+    return new ImmutableConciseSet(buf.asIntBuffer());
```
[the remark this answers is not in the document]
Comment (leventov) [reply processing/src/test/java/io/druid/collections/bitmap/BitmapBenchmark.java]: I removed one of the constructors, don't remember for what reason, maybe to reduce ambiguity. I don't see that one version is better than another.

File: processing/src/main/java/io/druid/collections/spatial/ImmutableRTree.java (hunk 2 above)
Comment (leventov) [reply processing/src/main/java/io/druid/collections/spatial/ImmutableRTree.java]: Direct ByteBuffers are not the thing that "magically makes everything faster". There are two main reasons to use direct ByteBuffers:
 - Reduce the size of Java Heap, (ordinary BBs are backed with `byte[]` arrays which are in the heap), to reduce GC pause times. Yes, even heap populated with arrays of primitives could affect GC pause times negatively. However, this comes with many downsides:
   - Native allocations are slower
   - Native allocations contribute to potentially unrecoverable fragmentation. On the contrary, Java Heap fragmentation is cleaned up, sooner or later.
   - You need ensure timely freeing of direct BBs, preferably with try-with-resources, not relying on reference cleaning

 - Avoid extra data copies when going native, particularly when a ByteBuffer participate network or disk IO calls (ByteChannel.read() / write())

Neither of this directly applies to `ImmutableRTree`.

**Comment by leventov** [comment]
@himanshug thanks. It should be said in `ColumnSerializer` javadoc, could you please add?

**Comment by himanshug** [comment]
@leventov please feel free to add the doc in this PR itself, not a major change to require new PR. It 
 should be along the lines of ....

```
ColumnSerializer can be implemented in custom aggregator extensions that would like to take full control of column serialization. That would implementation would be returned by overriding ComplexMetricsSerde.getSerializer(..).
```

File: processing/src/main/java/io/druid/segment/ColumnSelector.java [hunk 4]
```diff
@@ -31,5 +31,5 @@
   List<String> getColumnNames();
 
   @Nullable
-  Column getColumn(String columnName);
+  ColumnHolder getColumn(String columnName);
```
Comment (clintropolis) [inline processing/src/main/java/io/druid/segment/ColumnSelector.java]: Nit, maybe this should be called `getColumnHolder` or something to be a bit clearer since `ColumnHolder.getColumn` also exists and there are a handful of places that are effectively calling `selector.getColumn().getColumn()` which reads kind of funny.

Comment (leventov) [reply processing/src/main/java/io/druid/segment/ColumnSelector.java]: Thanks for noticing, renamed.

**Comment by leventov** [comment]
@nishantmonu51 please review this PR

File: processing/src/main/java/io/druid/segment/virtual/SingleStringInputDimensionSelector.java [hunk 5]
```diff
@@ -139,7 +139,7 @@ public String lookupName(final int id)
   @Override
   public boolean nameLookupPossibleInAdvance()
   {
-    return true;
+    return selector.nameLookupPossibleInAdvance();
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/virtual/SingleStringInputDimensionSelector.java]: This is not a bug, since the constructor verifies that `selector.nameLookupPossibleInAdvance()` returns true. (Otherwise, the optimization this class is trying to do is pointless.)

File: processing/src/main/java/io/druid/segment/virtual/SingleStringInputDimensionSelector.java [hunk 6]
```diff
@@ -153,7 +153,7 @@ public IdLookup idLookup()
   @Override
   public Object getObject()
   {
-    return defaultGetObject();
+    return lookupName(getRow().get(0));
```
Comment (gianm) [inline processing/src/main/java/io/druid/segment/virtual/SingleStringInputDimensionSelector.java]: I'm not sure this will work for empty rows in multi-value dimensions.

File: extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/quantiles/package-info.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+@EverythingIsNonnullByDefault
```
Comment (gianm) [inline extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/quantiles/package-info.java]: Is the quantiles package ready for this? (Everything nullable is annotated?)

Similar question for other packages. On most of the additions of EverythingIsNonnullByDefault I am wondering if the packages in question are ready. Right now, their parameters and return values are all treated as potentially nullable, which seems like the 'safe' bet if the annotations aren't there yet.

File: processing/src/main/java/io/druid/query/groupby/package-info.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+@EverythingIsNonnullByDefault
```
Comment (gianm) [inline processing/src/main/java/io/druid/query/groupby/package-info.java]: Is the groupBy package ready for this? (Everything nullable is annotated?)

File: benchmarks/src/main/java/io/druid/benchmark/package-info.java [hunk 9]
```diff
@@ -17,17 +17,7 @@
...
+@EverythingIsNonnullByDefault
+package io.druid.benchmark;
 
-import io.druid.guice.annotations.ExtensionPoint;
-import io.druid.segment.serde.Serializer;
-
-import java.io.IOException;
-
-@ExtensionPoint
-public interface GenericColumnSerializer extends Serializer
```
Comment (gianm) [inline benchmarks/src/main/java/io/druid/benchmark/package-info.java]: This was marked as an extension point since it's a return value of ComplexMetricSerde's `getSerializer` method. So I think we should revert this change. I also am not sure it's worth making even in the next major release, since there isn't a huge benefit and it would mean people need to adjust their extensions.

File: processing/src/main/java/io/druid/segment/virtual/SingleStringInputDimensionSelector.java (hunk 5 above)
Comment (leventov) [reply processing/src/main/java/io/druid/segment/virtual/SingleStringInputDimensionSelector.java]: Thanks, fixed

File: processing/src/main/java/io/druid/segment/virtual/SingleStringInputDimensionSelector.java (hunk 6 above)
Comment (leventov) [reply processing/src/main/java/io/druid/segment/virtual/SingleStringInputDimensionSelector.java]: Right, fixed. Thanks.

File: extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/quantiles/package-info.java (hunk 7 above)
Comment (leventov) [reply extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/quantiles/package-info.java]: Without this annotation, IntelliJ treats all method parameters that are not explicitly annotated `@Nullable` as "unknown" (*not* nullable!) (equivalent of [platform types](https://kotlinlang.org/docs/reference/java-interop.html#null-safety-and-platform-types) in Kotlin), and it doesn't warn you about anything, whether you treat them as nullable or not on the caller side or on the method body side.

So without `@EverythingIsNonnullByDefault`, there are two types of method parameters in the package: nullable (those that are annotated `@Nullable`), and "unknown" (those that are not annotated). You could also make parameters non-nullable by explicitly annotating them `@Nonnull`, but we never do this by convention.

With `@EverythingIsNonnullByDefault`, all "unknown" parameters are considered to be non-nullable, and IntelliJ starts to warn you about
 1) Nullable treatment on the method implementation side
 2) Passing provably nullable objects as the corresponding argument on the method calling side
 3) When the method overrides a method in a supertype, where the corresponding parameter is annotated `@Nullable`.

All those warnings are often false-positives, when the parameter is, in fact, nullable, but just wasn't annotated `@Nullable`. This was IntelliJ pushes you to propagate usage of `@Nullable` in the codebase.

So in fact according to our conventions all packages should be annotated `@EverythingIsNonnullByDefault`, I just never took time to do this in a separate PR.

File: benchmarks/src/main/java/io/druid/benchmark/package-info.java (hunk 9 above)
Comment (leventov) [reply benchmarks/src/main/java/io/druid/benchmark/package-info.java]: Next release is a major release (0.12.x -> 0.13). I don't think we should fall into conservatism (not fixing things because users may depend on them) until Druid 1.0, and clean up APIs eagerly until before that.

Comment (gianm) [reply benchmarks/src/main/java/io/druid/benchmark/package-info.java]: IMO it is painful for users to have to make small adjustments for non-functionality-affecting changes each time they upgrade to a new major Druid release. So I think it would be friendlier to do a 'big bang' and make any desired non-functionality-affecting, yet still backwards incompatible, changes like this all in one major release. Probably the one immediately prior to 1.0.

Comment (leventov) [reply benchmarks/src/main/java/io/druid/benchmark/package-info.java]: Ok. #6103.

File: processing/src/main/java/io/druid/query/aggregation/AggregatorUtil.java [hunk 10]
```diff
@@ -152,14 +154,14 @@
...
   }
 
-  public static BaseFloatColumnValueSelector makeColumnValueSelectorWithFloatDefault(
+  static BaseFloatColumnValueSelector makeColumnValueSelectorWithFloatDefault(
       final ColumnSelectorFactory metricFactory,
       final ExprMacroTable macroTable,
-      final String fieldName,
-      final String fieldExpression,
+      @Nullable final String fieldName,
+      @Nullable final String fieldExpression,
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/AggregatorUtil.java]: Add a comment that only one of fieldName or fieldExpression can be null.

File: processing/src/main/java/io/druid/query/aggregation/AggregatorUtil.java [hunk 11]
```diff
@@ -198,11 +200,11 @@ public boolean isNull()
...
   }
 
-  public static BaseLongColumnValueSelector makeColumnValueSelectorWithLongDefault(
+  static BaseLongColumnValueSelector makeColumnValueSelectorWithLongDefault(
       final ColumnSelectorFactory metricFactory,
       final ExprMacroTable macroTable,
-      final String fieldName,
-      final String fieldExpression,
+      @Nullable final String fieldName,
+      @Nullable final String fieldExpression,
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/AggregatorUtil.java]: Add a comment that only one of fieldName or fieldExpression can be null.

File: processing/src/main/java/io/druid/query/aggregation/AggregatorUtil.java [hunk 12]
```diff
@@ -239,11 +241,11 @@ public boolean isNull()
...
   }
 
-  public static BaseDoubleColumnValueSelector makeColumnValueSelectorWithDoubleDefault(
+  static BaseDoubleColumnValueSelector makeColumnValueSelectorWithDoubleDefault(
       final ColumnSelectorFactory metricFactory,
       final ExprMacroTable macroTable,
-      final String fieldName,
-      final String fieldExpression,
+      @Nullable final String fieldName,
+      @Nullable final String fieldExpression,
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/AggregatorUtil.java]: Add a comment that only one of fieldName or fieldExpression can be null.

File: processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java [hunk 13]
```diff
@@ -339,11 +340,11 @@ public double aggregate(final double current, final BaseObjectColumnValueSelecto
...
-              final Object[] arrayAsObjectArray = new Object[Array.getLength(arg)];
-              for (int j = 0; j < Array.getLength(arg); j++) {
-                arrayAsObjectArray[j] = Array.get(arg, j);
-              }
-              args[i + 1] = cx.newArray(scope, arrayAsObjectArray);
+              args[i + 1] = cx.newArray(scope, arrayToObjectArray(arg));
+            } else if (arg instanceof List) {
+              // Using toArray(Object[]), instead of just toArray(), because Arrays.asList()'s impl and similar List
+              // impls could clone the underlying array in toArray(), that could be not Object[], but e. g. String[].
+              args[i + 1] = cx.newArray(scope, ((List) arg).toArray(ObjectArrays.EMPTY_ARRAY));
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java]: do we have a test case for this ?

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java [hunk 14]
```diff
@@ -49,15 +47,11 @@
 
 public class DoubleFirstAggregatorFactory extends NullableAggregatorFactory<ColumnValueSelector>
 {
-  public static final Comparator VALUE_COMPARATOR = (o1, o2) -> Doubles.compare(
-      ((SerializablePair<Long, Double>) o1).rhs,
-      ((SerializablePair<Long, Double>) o2).rhs
-  );
+  public static final Comparator VALUE_COMPARATOR =
+      Comparator.comparingDouble(o -> ((SerializablePair<Long, Long>) o).rhs);
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: should this be SerializablePair<Long, Double> instead of SerializablePair<Long,Long> ?

File: processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java [hunk 15]
```diff
@@ -49,15 +47,11 @@
...
 
-  public static final Comparator TIME_COMPARATOR = (o1, o2) -> Longs.compare(
-      ((SerializablePair<Long, Object>) o1).lhs,
-      ((SerializablePair<Long, Object>) o2).lhs
-  );
+  public static final Comparator TIME_COMPARATOR =
+      Comparator.comparingLong(o -> ((SerializablePair<Long, Long>) o).lhs);
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/aggregation/first/DoubleFirstAggregatorFactory.java]: should this be SerializablePair<Long, Object> instead of SerializablePair<Long,Long> ?

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java [hunk 16]
```diff
@@ -329,14 +329,12 @@ public RowIterator(GroupByQuery query, final Cursor cursor, ByteBuffer metricsBu
...
-            throw new UnsupportedOperationException(
-                "GroupBy v1 does not support dimension selectors with unknown cardinality.");
-          }
-          dimensions.add(selector);
-          dimNames.add(dimSpec.getOutputName());
+        if (selector.getValueCardinality() == DimensionSelector.CARDINALITY_UNKNOWN) {
+          throw new UnsupportedOperationException(
+              "GroupBy v1 does not support dimension selectors with unknown cardinality.");
         }
+        dimensions.add(selector);
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: Note to self: reviewed till here.

File: extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/quantiles/package-info.java (hunk 7 above)
Comment (gianm) [reply extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/quantiles/package-info.java]: Is there an easier way to get this behavior than adding a hundred `package-info.java` files?

Comment (leventov) [reply extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/quantiles/package-info.java]: Good question. There is no such configuration in IntelliJ currently as far as I can tell. https://youtrack.jetbrains.com/issue/IDEA-197110

File: processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java (hunk 13 above)
Comment (leventov) [reply processing/src/main/java/io/druid/query/aggregation/JavaScriptAggregatorFactory.java]: This change is going to be required in the following PR, where I replace some uses of Java arrays with lists. It's going to be tested then.

**Comment by leventov** [comment]
@nishantmonu51 could you continue review?

File: processing/src/main/java/io/druid/segment/QueryableIndexColumnSelectorFactory.java [hunk 17]
```diff
@@ -79,28 +78,34 @@ private DimensionSelector makeDimensionSelectorUndecorated(DimensionSpec dimensi
...
 
-    @SuppressWarnings("unchecked")
-    DictionaryEncodedColumn<String> column = (DictionaryEncodedColumn<String>)
-        columnCache.computeIfAbsent(dimension, d -> closer.register(columnDesc.getDictionaryEncoding()));
+    BaseColumn column = columnCache.computeIfAbsent(dimension, d -> {
+      BaseColumn col = columnHolder.getColumn();
+      if (col instanceof DictionaryEncodedColumn) {
+        return closer.register(col);
+      } else {
+        return null;
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/segment/QueryableIndexColumnSelectorFactory.java]: should this return NilColumnValueSelector.instance() instead ? 
If this is expected, need to make sure the method DimensionSpec.decorate accept nulls and makeDimensionSelector is ok to return nulls.

File: processing/src/main/java/io/druid/segment/StorageAdapter.java [hunk 18]
```diff
@@ -39,12 +39,8 @@
   Iterable<String> getAvailableMetrics();
 
   /**
-   * Returns the number of distinct values for the given dimension column
-   * For dimensions of unknown cardinality, e.g. __time this currently returns
-   * Integer.MAX_VALUE
-   *
-   * @param column
-   * @return
+   * Returns the number of distinct values for the given column if known, or {@link Integer#MAX_VALUE} is unknown or
```
Comment (nishantmonu51) [inline processing/src/main/java/io/druid/segment/StorageAdapter.java]: typo "is unknown" -> "if unknown"

File: processing/src/main/java/io/druid/segment/QueryableIndexColumnSelectorFactory.java (hunk 17 above)
Comment (leventov) [reply processing/src/main/java/io/druid/segment/QueryableIndexColumnSelectorFactory.java]: This is return from the lambda in `computeIfAbsent()`, not the outer method. Added comment for clarity.

File: processing/src/main/java/io/druid/segment/StorageAdapter.java (hunk 18 above)
Comment (leventov) [reply processing/src/main/java/io/druid/segment/StorageAdapter.java]: Thanks, fixed.

**Comment by leventov** [comment]
@nishantmonu51 thanks for review. I've occasionally run Optimize Imports (well, I wanted to do this in IntelliJ. But I thought it will run only on the affected files, not on the whole codebase.) so now the change seems to be much bigger than it used to be. But all extra changes are just mechanical import changes.

**Comment by leventov** [comment]
@nishantmonu51 the PR passed Travis. Last 11 commits are about combating unrelated transient test failures.

**Comment by leventov** [comment]
@nishantmonu51 could you please take a look?

**Comment by leventov** [comment]
@nishantmonu51 @gianm could we finish off this PR? Given it effectively has three approvals for three weeks, and still not merged?

**Comment by gianm** [comment]
I haven't had a chance to review the entire thing yet, beyond the partial review I did in https://github.com/apache/incubator-druid/pull/5957#pullrequestreview-142270140. Please don't consider me as blocking this PR, just not yet having the opportunity to review it in full.

By way of explanation as to why: I can't speak for everyone, but personally I am prioritizing my time towards reviewing PRs that make functional changes (which we are getting a lot of!) vs. ones that are refactorings/cleanups. Doesn't mean I'll never get to it, but hopefully that explains why I haven't yet. I understand this is a dev blocker for you, so I guess all I can say to that is, I would suggest avoiding making functional changes dependent on large nonfunctional ones.

**Comment by fjy** [comment]
I think this PR should be looked at in 0.13.1

**Comment by fjy** [comment]
I think the PR should be broken into multiple PRs to make it much easier to review. The impact of breaking changes in such a significant PR impacting so many files is high. Furthermore, I'd like to see each separate PR comprehensively describe what was changed, and what the benefit was. I'm moving the milestone to 0.13.1 for now so give us enough time to properly review this PR.

**Comment by gianm** [comment]
Is this still "Incompatible"?

**Comment by gianm** [comment]
The test failures look legit (although are not a big deal) and are related to the renaming of HLLCV1. I don't think it's important that the value stay the same, so I think updating the tests is fine.

```
Failed tests: 
org.apache.druid.sql.calcite.http.SqlResourceTest.testCsvResultFormat(org.apache.druid.sql.calcite.http.SqlResourceTest)
  Run 1: SqlResourceTest.testCsvResultFormat:500 expected:<[2000-01-01T00:00:00.000Z,1,,a,1.0,1.0,org.apache.druid.hll.HLLCV1,, 2000-01-02T00:00:00.000Z,1,10.1,,2.0,2.0,org.apache.druid.hll.HLLCV1,, , ]> but was:<[2000-01-01T00:00:00.000Z,1,,a,1.0,1.0,org.apache.druid.hll.VersionOneHyperLogLogCollector,, 2000-01-02T00:00:00.000Z,1,10.1,,2.0,2.0,org.apache.druid.hll.VersionOneHyperLogLogCollector,, , ]>
  Run 2: SqlResourceTest.testCsvResultFormat:500 expected:<[2000-01-01T00:00:00.000Z,1,,a,1.0,1.0,org.apache.druid.hll.HLLCV1,, 2000-01-02T00:00:00.000Z,1,10.1,,2.0,2.0,org.apache.druid.hll.HLLCV1,, , ]> but was:<[2000-01-01T00:00:00.000Z,1,,a,1.0,1.0,org.apache.druid.hll.VersionOneHyperLogLogCollector,, 2000-01-02T00:00:00.000Z,1,10.1,,2.0,2.0,org.apache.druid.hll.VersionOneHyperLogLogCollector,, , ]>
  Run 3: SqlResourceTest.testCsvResultFormat:500 expected:<[2000-01-01T00:00:00.000Z,1,,a,1.0,1.0,org.apache.druid.hll.HLLCV1,, 2000-01-02T00:00:00.000Z,1,10.1,,2.0,2.0,org.apache.druid.hll.HLLCV1,, , ]> but was:<[2000-01-01T00:00:00.000Z,1,,a,1.0,1.0,org.apache.druid.hll.VersionOneHyperLogLogCollector,, 2000-01-02T00:00:00.000Z,1,10.1,,2.0,2.0,org.apache.druid.hll.VersionOneHyperLogLogCollector,, , ]>
  Run 4: SqlResourceTest.testCsvResultFormat:500 expected:<[2000-01-01T00:00:00.000Z,1,,a,1.0,1.0,org.apache.druid.hll.HLLCV1,, 2000-01-02T00:00:00.000Z,1,10.1,,2.0,2.0,org.apache.druid.hll.HLLCV1,, , ]> but was:<[2000-01-01T00:00:00.000Z,1,,a,1.0,1.0,org.apache.druid.hll.VersionOneHyperLogLogCollector,, 2000-01-02T00:00:00.000Z,1,10.1,,2.0,2.0,org.apache.druid.hll.VersionOneHyperLogLogCollector,, , ]>
```

**Comment by leventov** [comment]
@gianm thanks a lot for review and the suggestions - I'll account in future PRs.
